xgen: init at 0-unstable-2023-07-02

imalsogreg opened the pull request:
## Description of changes

Adds support for [xgen](https://github.com/xuri/xgen), a tool that generates Go/C/Java/Rust/TypeScript code from an .xsd (XML Schema Description) file.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

imalsogreg committed:
xgen: init at 0-unstable-2023-07-02

isabelroses commented on pkgs/development/tools/xgen/default.nix:
```diff
@@ -0,0 +1,28 @@
+
+  src = fetchFromGitHub {
+    owner = "xuri";
+    repo = pname;
```
`repo = pname` is an anti pattern.

imalsogreg replied:
Fixed, thank you.

isabelroses reviewed (changes requested):
Please move the package to the `by-name` standard, to see more read here https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md

imalsogreg commented:
@isabelroses Thank you for the pointer about by-name!

imalsogreg commented:
@isabelroses can we merge? :)

isabelroses commented on pkgs/by-name/xg/xgen/package.nix:
```diff
@@ -0,0 +1,28 @@
+
+  # This is a pseudoversion including the fetch date and git hash.
+  # The upstream project does not provide official releases or version tags.
+  version = "v0.0.0-20240523172100-db840e1a460537cf4f90c2a88cffe11ef354dd9f";
```
> Example: Given a project that has no tags / released versions at all, or applies versionless tags like latest or YYYY-MM-DD-Build, a commit authored on March 15, 2022 would have version = "0-unstable-2022-03-15".

https://github.com/NixOS/nixpkgs/blob/77874773a16d1c0507ee6a37c83a2da10bc2a276/pkgs/README.md?plain=1#L357

isabelroses commented:
> @isabelroses can we merge? :)

I don't have those kinds of permissions sadly. But I'll raise this when the latest comment is addressed.

isabelroses commented:
Also please squash your commits down into `xgen: init at 0-unstable-2023-07-02` and change the PR name to the same.

imalsogreg commented:
@isabelroses 
 - squashed to 1 commit
 - renamed commit to `xgen: init at 0-unstable-2023-07-02`
 - renamed PR to `xgen: init at 0-unstable-2023-07-02`

Thank you for pointing me to these conventions!

slashformotion commented:
Hello, 
I am using this program and currently using the golang toolchain to get it on my machine, it would be great to get this merged so that we can put it in our nixos and nix-darwin configs at work. 

Many thanks to all of you for this work.

Pandapip1 commented on pkgs/by-name/xg/xgen/package.nix:
```diff
@@ -0,0 +1,28 @@
+buildGoModule rec {
+  pname = "xgen";
+
+  # This is a pseudoversion including the fetch date and git hash.
+  # The upstream project does not provide official releases or version tags.
+  version = "v0-unstable-2023-07-02";
```
```suggestion
  version = "0-unstable-2023-07-02";
```

The `0-unstable-` versioning is well-established and does not require explaining. Additionally, according to the naming guidelines, the `version` MUST start with a number. You cannot prefix said number with a `v`.

Pandapip1 commented on pkgs/by-name/xg/xgen/package.nix:
```diff
@@ -0,0 +1,28 @@
+
+  vendorHash = "sha256-YU/iQnZs4/oUiJw3syW1jxEpeQLyOMElsjzWCNsucds=";
+
+  meta = with lib; {
+    description = "A XSD to Go/C/Java/Rust/TypeScript code generator";
+    homepage = "https://github.com/xuri/xgen";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ imalsogreg ];
+    mainProgram = "xgen";
+  };
```
```suggestion
  meta = {
    description = "A XSD to Go/C/Java/Rust/TypeScript code generator";
    homepage = "https://github.com/xuri/xgen";
    license = lib.licenses.bsd3;
    maintainers = with lib.maintainers; [ imalsogreg ];
    mainProgram = "xgen";
  };
```

`meta = with lib; {` is discouraged.
